### PR TITLE
Split graph

### LINKED ISSUE #53: split graph

### PR DESCRIPTION
**Pull request by LucaStaus**
Adds functions for creating a split graph, calculating the splittance of a graph and checking if a graph is a split graph.

### CODE REVIEW

File: src/main/kotlin/sipura/generate/Factory.kt [hunk 1]
```diff
@@ -78,4 +78,36 @@ object Factory {
...
+     * possible edges between the independent set and the clique.
+     *
+     * The vertices of the independent set are [1, ..., [sizeIS]] and the vertices of the clique
+     * are [[sizeIS] + 1, ..., [sizeIS] + [sizeC]].
+     *
+     * @throws IllegalArgumentException if either [sizeIS] or [sizeC] are not positive.
+     * @return The created split graph.
+     */
+    fun createSplitGraph(sizeIS: Int, sizeC: Int): SimpleGraph<Int> {
+        if (sizeIS < 0 || sizeC < 0) throw IllegalArgumentException("Both sides must have non-negative size")
```
Comment (moritz-gross) [inline src/main/kotlin/sipura/generate/Factory.kt]: 0 is not positive, but no exception is thrown for it.
We have to make this consistent, even though I'm not sure whether 0 should be a valid input.

Comment (moritz-gross) [reply src/main/kotlin/sipura/generate/Factory.kt]: Ok from the unit tests it seems like size 0 is fine for both parameters. IMO we can keep it like that.

Then I'd just reword it from "not positive" to "negative" to exclude 0.

Comment (LucaStaus) [reply src/main/kotlin/sipura/generate/Factory.kt]: done

File: src/main/kotlin/sipura/generate/Factory.kt [hunk 2]
```diff
@@ -78,4 +78,36 @@ object Factory {
...
+            for (v2 in (v1 + 1)..(sizeIS + sizeC)) {
+                g.addEdge(v1, v2)
+            }
+        }
+        // add all edges between IS and Clique
+        for (v1 in 1..sizeIS) {
+            for (v2 in (sizeIS + 1)..(sizeIS + sizeC)) {
+                g.addEdge(v1, v2)
+            }
+        }
```
Comment (moritz-gross) [inline src/main/kotlin/sipura/generate/Factory.kt:110]: Could we do this more concise with your 'disjointUnion' function?

File: src/main/kotlin/sipura/alg/GraphProperty.kt [hunk 3]
```diff
@@ -219,4 +219,58 @@ object GraphProperty {
...
+     *
+     * Runtime: O(n + m)
+     * @return the splittance of [g].
+     */
+    fun <V> minSplittance(g: SimpleGraph<V>): Int {
+        if (g.n == 0) return 0 // special case because maxDegree() throws exception for size 0
+        val maxDegree = g.maxDegree()
+        var degreeSumTotal = 0
+        // order vertices by degree in linear time
+        val degreeOrdering = Array<LinkedList<V>>(maxDegree + 1) { LinkedList() }
```
Comment (moritz-gross) [inline src/main/kotlin/sipura/alg/GraphProperty.kt:238]: create new function for degreeOrdering?
Seems like something that can be reused well.

File: src/main/kotlin/sipura/alg/GraphProperty.kt [hunk 4]
```diff
@@ -219,4 +219,58 @@ object GraphProperty {
...
+
+    /**
+     * Calculates the splittance of the given graph [g].
+     *
+     * The splittance of [g] is the smallest number of edge deletions or edge insertions that are required to
+     * transform [g] into a split graph. That means the splittance of [g] is zero if and only if [g] is a split graph.
+     *
+     * Runtime: O(n + m)
+     * @return the splittance of [g].
+     */
```
Comment (moritz-gross) [inline src/main/kotlin/sipura/alg/GraphProperty.kt]: reference https://en.wikipedia.org/wiki/Splittance#Calculation_from_degree_sequence?
I sadly didn't find an arxiv-link or something similar, just this springer link with a paywall: https://link.springer.com/article/10.1007/BF02579333.

File: src/main/kotlin/sipura/generate/Factory.kt (hunk 2 above)
Comment (LucaStaus) [reply src/main/kotlin/sipura/generate/Factory.kt:110]: The code is already written and it works. So might aswell keep it.

File: src/main/kotlin/sipura/alg/GraphProperty.kt (hunk 3 above)
Comment (LucaStaus) [reply src/main/kotlin/sipura/alg/GraphProperty.kt:238]: I created an Issue for that: https://github.com/sipura/Sipura/issues/67

File: src/main/kotlin/sipura/alg/GraphProperty.kt (hunk 4 above)
Comment (LucaStaus) [reply src/main/kotlin/sipura/alg/GraphProperty.kt]: I added the link.

**Pull request merged**